EIPS/eip-custom_errors_for_erc_tokens.md

```diff
@@ -0,0 +1,504 @@
+---
+eip: [assignedEIPNumber]
+title: Custom errors for ERC tokens
+description: List of Solidity custom errors covering standard token implementations (EIP-20, EIP-721 and EIP-1155)
+author: Ernesto García (@ernestognw), Francisco Giordano (@frangio), Hadrien Croubois (@Amxx)
+discussions-to: <URL>
+status: Draft
+type: Standards Track
+category: ERC
+requires: 20, 721, 1155
+created: <date created on, in ISO 8601 (yyyy-mm-dd) format>
+---
+
+## Abstract
+
+The following specification allows for the use of a standard list of Solidity [custom errors](https://blog.soliditylang.org/2021/04/21/custom-errors/) to be used within [EIP-20](./eip-20.md), [EIP-721](./eip-721.md) and [EIP-1155](./eip-1155.md) tokens.
+
+Ethereum applications (DApps) and wallets have historically used [revert statements](https://docs.soliditylang.org/en/v0.8.16/control-structures.html#revert-statement) to show relevant failure data to the users, however, this EIP provides a standard list of errors designed to give at least the same relevant information, but in a structured and expected way.
+
+## Motivation
+
+Since the introduction of Solidity custom errors in [v0.8.4](https://github.com/ethereum/solidity/releases/tag/v0.8.4), these have provided a way to show failures in a more expresive way with dynamic arguments, while reducing deployment costs.
+
+At the moment of the release of custom errors, the standard tokens ([EIP-20](./eip-20.md), [EIP-721](./eip-721.md), [EIP-1155](./eip-1155.md)) were already in finalized state, so no error specification was included.
+
+An error specification will allow users to expect more consistent error messages across applications or testing environments, while exposing pertinent arguments and overall reducing the need of writing expensive revert strings in the deployment bytecode.
+
+The list of token standard errors is defined in the specification below.
+
+## Specification
+
+The key words "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT", "SHOULD", "SHOULD NOT", "RECOMMENDED", "NOT RECOMMENDED", "MAY", and "OPTIONAL" in this document are to be interpreted as described in RFC 2119 and RFC 8174.
+
+The provided interfaces with errors SHOULD conform with the [error grammar rules](#error-grammar-rules) described in the [rationale](#rationale) section.
+
+It's important to note that errors MAY be thrown in each scenario, but that's up to the implementers, unless a revert reason is specified in their corresponding EIPs.
+
+This EIP aims to provide a clear standard only in case an error will be thrown.
+
+### [EIP-20](./eip-20.md)
+
+#### `ERC20InvalidSender(address)`
+
+Indicates a failure with the token sender.
+
+##### Parameters
+
+- `address _sender`: The address of whose tokens are transferred from.
```

Comment:
IMO parameters are mostly self-explanatory and I would remove this section. If we need to explain their meaning, I would add a small section at the beginning of the EIP-20 section saying "here is what we mean when we say sender, receiver, spender, etc", and we can also say what we mean by the "needed" parameter in general.

Reply:
> IMO parameters are mostly self-explanatory
Agree.

But there's something that's feeling off for me here.

Adding a section where we define subjects makes me feel that we should define the list of subjects for this EIP, which is something we previously stated not to do. Still, `_needed` is not a subject so I'm not sure my argument is completely correct.

I'll think about it but I'm sharing in case you have something to add :)

Reply:
>something we previously stated not to do

I didn't see them as belonging in the rationale section, but I think it would make sense in the specification section as documentation.

Reply:
I just created a glossary table. I'd think it's expressive enough but something feels weird. I'd appreciate help with the wording.